x/leverage/keeper/keeper.go

```diff
@@ -31,147 +31,163 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+		return err
+	}
+
+	return nil
 }
 
-// WithdrawAsset attempts to deposit uTokens into the leverage module in exchange for original assets.
-// If utoken type is invalid or account balance insufficient on either side, does nothing and returns false.
-// TODO: Panic if partially executed then fail?
-func (k Keeper) WithdrawAsset(ctx sdk.Context, msg types.MsgWithdrawAsset) bool {
-	lender, err := sdk.AccAddressFromBech32(msg.GetLender())
+// WithdrawAsset attempts to deposit uTokens into the leverage module in exchange
+// for the original tokens lent. If the uToken type is invalid or account balance
+// insufficient on either side, we return an error.
+func (k Keeper) WithdrawAsset(ctx sdk.Context, msg types.MsgWithdrawAsset) error {
+	lenderAddr, err := sdk.AccAddressFromBech32(msg.GetLender())
 	if err != nil {
-		// Could not unmarshal address (Bech32)
-		return false
+		return err
 	}
-	utokens := msg.GetAmount()
-	if !k.IsAcceptedUtoken(ctx, utokens.Denom) {
-		// Not an accepted utoken type
-		return false
+
+	uToken := msg.GetAmount()
+	if !k.IsAcceptedUToken(ctx, uToken.Denom) {
+		return sdkerrors.Wrap(types.ErrInvalidAsset, uToken.String())
 	}
-	assetDenom := k.ToBaseAssetDenom(ctx, utokens.Denom)
-	if !k.bankKeeper.HasBalance(ctx, lender, utokens) {
+
+	if !k.bankKeeper.HasBalance(ctx, lenderAddr, uToken) {
 		// Lender does not have the uTokens they intend to redeem
-		return false
-	}
-	assets := sdk.NewCoin(assetDenom, utokens.Amount) // TODO: Use exchange rate instead of 1:1 redeeming
-	/*
-		TODO: once we have leverage module's account address
-		if k.bankKeeper.HasBalance(ctx,module_account_address,assets) == false {
-			return false
-		}
-	*/
-	// TODO: Add additional checks to be sure tokens/accounts are good to send from before doing the
-	// irreversible
-	if k.bankKeeper.SendCoinsFromAccountToModule(ctx, lender, "leverage", sdk.Coins{utokens}) != nil {
-		// Error depositing utokens
-		return false
-	}
-	if k.bankKeeper.SendCoinsFromModuleToAccount(ctx, "leverage", lender, sdk.Coins{assets}) != nil {
-		// Error releasing assets
-		// TODO: Either make this a panic, or reverse the sending of uTokens
-		return false
-	}
-	if k.bankKeeper.BurnCoins(ctx, "leverage", sdk.Coins{utokens}) != nil {
-		// Error burning utokens
-		// TODO: Either make this a panic, or reverse the depositing or utokens and releasing of assets
-		return false
-	}
-	return false
+		return sdkerrors.Wrap(types.ErrInsufficientBalance, uToken.String())
+	}
+
+	// ensure the tokens exist in the leverage module account's balance
+	// TODO: Use exchange rate instead of 1:1 redeeming
+	tokenDenom := k.FromUTokenToTokenDenom(ctx, uToken.Denom)
+	token := sdk.NewCoin(tokenDenom, uToken.Amount)
+	if !k.bankKeeper.HasBalance(ctx, authtypes.NewModuleAddress(types.ModuleName), token) {
+		// TODO: We should never enter this case -- considering a panic
+		return sdkerrors.Wrap(types.ErrInsufficientBalance, uToken.String())
```

Comment:
```suggestion
		return sdkerrors.Wrap(types.ErrInsufficientBalance, token.String())
```
The balance that's insufficient here is the token on the module's side, so it might be better to string that one into the error instead of the uToken

Reply:
Good catch!